Looking at the test failure, I think there is some bug in tests class loading @holly-cummins :

```
2025-11-25T17:56:27.6369691Z [ERROR] io.quarkus.it.keycloak.CodeFlowAuthorizationTest -- Time elapsed: 1.235 s <<< ERROR!
2025-11-25T17:56:27.6370434Z java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
2025-11-25T17:56:27.6371246Z 	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:663)
2025-11-25T17:56:27.6372195Z 	at io.quarkus.test.junit.QuarkusTestExtension.interceptBeforeAllMethod(QuarkusTestExtension.java:730)
2025-11-25T17:56:27.6372921Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
2025-11-25T17:56:27.6373422Z Caused by: java.lang.RuntimeException: Failed to start quarkus
2025-11-25T17:56:27.6373931Z 	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
2025-11-25T17:56:27.6374435Z 	at io.quarkus.runtime.Application.start(Application.java:101)
2025-11-25T17:56:27.6374936Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:586)
2025-11-25T17:56:27.6375681Z 	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:350)
2025-11-25T17:56:27.6376412Z 	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:229)
2025-11-25T17:56:27.6377203Z 	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:637)
2025-11-25T17:56:27.6377981Z 	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:682)
2025-11-25T17:56:27.6378644Z 	... 1 more
2025-11-25T17:56:27.6379645Z Caused by: java.lang.RuntimeException: Internal error: shutdownContext is null. This probably happened because Quarkus failed to start properly in an earlier step, or because tests were run on a Quarkus instance that had already been shut down.
2025-11-25T17:56:27.6381008Z 	at io.quarkus.runtime.configuration.ConfigRecorder.releaseConfig(ConfigRecorder.java:113)
2025-11-25T17:56:27.6381932Z 	at io.quarkus.runner.recorded.ConfigGenerationBuildStep$releaseConfigOnShutdown561040398.deploy_0(Unknown Source)
2025-11-25T17:56:27.6382947Z 	at io.quarkus.runner.recorded.ConfigGenerationBuildStep$releaseConfigOnShutdown561040398.deploy(Unknown Source)
2025-11-25T17:56:27.6383545Z 	... 8 more
2025-11-25T17:56:27.6383685Z 
2025-11-25T17:56:27.6463524Z 2025-11-25 17:56:27,645 ERROR [io.quarkus.runner.bootstrap.StartupActionImpl] (main) Internal error: Could not read configuration while evaluating whether to run class io.quarkus.it.keycloak.CodeFlowAuthorizationTest. Please let the Quarkus team know what you were doing when this error happened.
2025-11-25T17:56:27.6464959Z [INFO] Running io.quarkus.it.keycloak.CodeFlowAuthorizationTest
2025-11-25T17:56:27.6545645Z [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.001 s <<< FAILURE! -- in io.quarkus.it.keycloak.CodeFlowAuthorizationTest
2025-11-25T17:56:27.6546419Z [ERROR] io.quarkus.it.keycloak.CodeFlowAuthorizationTest -- Time elapsed: 0.001 s <<< ERROR!
2025-11-25T17:56:27.6547917Z org.junit.jupiter.engine.execution.ConditionEvaluationException: Failed to evaluate condition [io.quarkus.test.junit.QuarkusTestExtension]: Internal error: Test class was loaded with an unexpected classloader (QuarkusClassLoader:Quarkus Base Runtime ClassLoader: TEST for JUnitQuarkusTest-no-profile (QuarkusTest)@132fab82) or the thread context classloader (jdk.internal.loader.ClassLoaders$AppClassLoader@904a5a38) was incorrect.
2025-11-25T17:56:27.6549261Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
2025-11-25T17:56:27.6549731Z 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
2025-11-25T17:56:27.6550176Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
2025-11-25T17:56:27.6550619Z 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
2025-11-25T17:56:27.6551183Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
2025-11-25T17:56:27.6551641Z 	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1685)
2025-11-25T17:56:27.6552330Z 	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
2025-11-25T17:56:27.6553023Z 	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:540)
2025-11-25T17:56:27.6553667Z 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:526)
2025-11-25T17:56:27.6554153Z 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:512)
2025-11-25T17:56:27.6554620Z 	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
2025-11-25T17:56:27.6555038Z 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:239)
2025-11-25T17:56:27.6555716Z 	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
2025-11-25T17:56:27.6556114Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
2025-11-25T17:56:27.6557126Z Caused by: java.lang.IllegalStateException: Internal error: Test class was loaded with an unexpected classloader (QuarkusClassLoader:Quarkus Base Runtime ClassLoader: TEST for JUnitQuarkusTest-no-profile (QuarkusTest)@132fab82) or the thread context classloader (jdk.internal.loader.ClassLoaders$AppClassLoader@904a5a38) was incorrect.
2025-11-25T17:56:27.6558349Z 	at io.quarkus.test.junit.AbstractJvmQuarkusTestExtension.evaluateExecutionCondition(AbstractJvmQuarkusTestExtension.java:206)
2025-11-25T17:56:27.6558942Z 	... 14 more
2025-11-25T17:56:27.6559438Z Caused by: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
2025-11-25T17:56:27.6560147Z 	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:601)
2025-11-25T17:56:27.6560566Z 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1263)
2025-11-25T17:56:27.6561041Z 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1292)
2025-11-25T17:56:27.6561436Z 	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1328)
2025-11-25T17:56:27.6561761Z 	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1412)
2025-11-25T17:56:27.6562200Z 	at io.smallrye.config.SmallRyeConfigProviderResolver.getFactoryFor(SmallRyeConfigProviderResolver.java:102)
2025-11-25T17:56:27.6562757Z 	at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:78)
2025-11-25T17:56:27.6563300Z 	at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:66)
2025-11-25T17:56:27.6563788Z 	at org.eclipse.microprofile.config.ConfigProvider.getConfig(ConfigProvider.java:85)
2025-11-25T17:56:27.6564351Z 	at io.quarkus.test.junit.AbstractJvmQuarkusTestExtension.evaluateExecutionCondition(AbstractJvmQuarkusTestExtension.java:162)
2025-11-25T17:56:27.6564783Z 	... 14 more
2025-11-25T17:56:27.6564870Z 
2025-11-25T17:56:27.6580301Z 2025-11-25 17:56:27,657 ERROR [io.quarkus.runner.bootstrap.StartupActionImpl] (main) Internal error: Could not read configuration while evaluating whether to run class io.quarkus.it.keycloak.CodeFlowAuthorizationTest. Please let the Quarkus team know what you were doing when this error happened.
2025-11-25T17:56:27.6581782Z [INFO] Running io.quarkus.it.keycloak.CodeFlowAuthorizationTest
2025-11-25T17:56:27.6666244Z [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.001 s <<< FAILURE! -- in io.quarkus.it.keycloak.CodeFlowAuthorizationTest
2025-11-25T17:56:27.6667338Z [ERROR] io.quarkus.it.keycloak.CodeFlowAuthorizationTest -- Time elapsed: 0.001 s <<< ERROR!
2025-11-25T17:56:27.6669100Z org.junit.jupiter.engine.execution.ConditionEvaluationException: Failed to evaluate condition [io.quarkus.test.junit.QuarkusTestExtension]: Internal error: Test class was loaded with an unexpected classloader (QuarkusClassLoader:Quarkus Base Runtime ClassLoader: TEST for JUnitQuarkusTest-no-profile (QuarkusTest)@132fab82) or the thread context classloader (jdk.internal.loader.ClassLoaders$AppClassLoader@904a5a38) was incorrect.
2025-11-25T17:56:27.6670500Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
2025-11-25T17:56:27.6670945Z 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
2025-11-25T17:56:27.6671371Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
2025-11-25T17:56:27.6671819Z 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
2025-11-25T17:56:27.6672460Z 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
2025-11-25T17:56:27.6673189Z 	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1685)
2025-11-25T17:56:27.6673705Z 	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
2025-11-25T17:56:27.6674267Z 	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:540)
2025-11-25T17:56:27.6675658Z 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:526)
2025-11-25T17:56:27.6676412Z 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:512)
2025-11-25T17:56:27.6677107Z 	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
2025-11-25T17:56:27.6677767Z 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:239)
2025-11-25T17:56:27.6678589Z 	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
2025-11-25T17:56:27.6679174Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
2025-11-25T17:56:27.6680729Z Caused by: java.lang.IllegalStateException: Internal error: Test class was loaded with an unexpected classloader (QuarkusClassLoader:Quarkus Base Runtime ClassLoader: TEST for JUnitQuarkusTest-no-profile (QuarkusTest)@132fab82) or the thread context classloader (jdk.internal.loader.ClassLoaders$AppClassLoader@904a5a38) was incorrect.
2025-11-25T17:56:27.6682881Z 	at io.quarkus.test.junit.AbstractJvmQuarkusTestExtension.evaluateExecutionCondition(AbstractJvmQuarkusTestExtension.java:206)
2025-11-25T17:56:27.6683669Z 	... 14 more
2025-11-25T17:56:27.6684453Z Caused by: java.util.ServiceConfigurationError: io.smallrye.config.SmallRyeConfigFactory: io.quarkus.runtime.configuration.QuarkusConfigFactory not a subtype
2025-11-25T17:56:27.6685542Z 	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:601)
2025-11-25T17:56:27.6686251Z 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1263)
2025-11-25T17:56:27.6687107Z 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1292)
2025-11-25T17:56:27.6687749Z 	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1328)
2025-11-25T17:56:27.6688297Z 	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1412)
2025-11-25T17:56:27.6689067Z 	at io.smallrye.config.SmallRyeConfigProviderResolver.getFactoryFor(SmallRyeConfigProviderResolver.java:102)
2025-11-25T17:56:27.6689988Z 	at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:78)
2025-11-25T17:56:27.6690544Z 	at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:66)
2025-11-25T17:56:27.6691136Z 	at org.eclipse.microprofile.config.ConfigProvider.getConfig(ConfigProvider.java:85)
2025-11-25T17:56:27.6691807Z 	at io.quarkus.test.junit.AbstractJvmQuarkusTestExtension.evaluateExecutionCondition(AbstractJvmQuarkusTestExtension.java:162)
2025-11-25T17:56:27.6692323Z 	... 14 more
2025-11-25T17:56:27.6692414Z 
2025-11-25T17:56:27.7096785Z [INFO] 
2025-11-25T17:56:27.7097114Z [INFO] Results:
2025-11-25T17:56:27.7097413Z [INFO] 
2025-11-25T17:56:27.7097678Z [ERROR] Failures: 
2025-11-25T17:56:27.7098542Z [ERROR]   CodeFlowAuthorizationTest.testCodeFlowUserInfoCachedInIdToken:431 expected: <true> but was: <false>
2025-11-25T17:56:27.7099466Z [ERROR] Errors: 
2025-11-25T17:56:27.7100372Z [ERROR] io.quarkus.it.keycloak.CodeFlowAuthorizationTest.null
2025-11-25T17:56:27.7101284Z [ERROR]   Run 1: CodeFlowAuthorizationTest » Runtime java.lang.RuntimeException: Failed to start quarkus
2025-11-25T17:56:27.7103957Z [ERROR]   Run 2: CodeFlowAuthorizationTest » ConditionEvaluation Failed to evaluate condition [io.quarkus.test.junit.QuarkusTestExtension]: Internal error: Test class was loaded with an unexpected classloader (QuarkusClassLoader:Quarkus Base Runtime ClassLoader: TEST for JUnitQuarkusTest-no-profile (QuarkusTest)@132fab82) or the thread context classloader (jdk.internal.loader.ClassLoaders$AppClassLoader@904a5a38) was incorrect.
2025-11-25T17:56:27.7108307Z [ERROR]   Run 3: CodeFlowAuthorizationTest » ConditionEvaluation Failed to evaluate condition [io.quarkus.test.junit.QuarkusTestExtension]: Internal error: Test class was loaded with an unexpected classloader (QuarkusClassLoader:Quarkus Base Runtime ClassLoader: TEST for JUnitQuarkusTest-no-profile (QuarkusTest)@132fab82) or the thread context classloader (jdk.internal.loader.ClassLoaders$AppClassLoader@904a5a38) was incorrect.
2025-11-25T17:56:27.7110514Z [INFO] 
```

--- a/docs/src/main/asciidoc/security-oidc-expanded-configuration.adoc
+++ b/docs/src/main/asciidoc/security-oidc-expanded-configuration.adoc
@@ -1310,6 +1310,7 @@ In turn, a client that requested the protected resource metadata can use its pro
 
 |quarkus.oidc.resource-metadata.enabled |false|Enable resource metadata properties
 |quarkus.oidc.resource-metadata.resource ||Resource identifier
+|quarkus.oidc.resource-metadata.scopes ||Supported scopes
 |quarkus.oidc.resource-metadata.authorization-server ||Authorization server URL
 |quarkus.oidc.resource-metadata.force-https-scheme |true|Force that a resource identifier URL has an HTTPS scheme
 |====

--- a/extensions/oidc-common/runtime/src/main/java/io/quarkus/oidc/common/runtime/OidcConstants.java
+++ b/extensions/oidc-common/runtime/src/main/java/io/quarkus/oidc/common/runtime/OidcConstants.java
@@ -112,5 +112,6 @@ public final class OidcConstants {
 
     public static final String RESOURCE_METADATA_WELL_KNOWN_PATH = "/.well-known/oauth-protected-resource";
     public static final String RESOURCE_METADATA_RESOURCE = "resource";
+    public static final String RESOURCE_METADATA_SCOPES = "scopes_supported";
     public static final String RESOURCE_METADATA_AUTHORIZATION_SERVERS = "authorization_servers";
 }

--- a/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java
+++ b/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java
@@ -2759,6 +2759,7 @@ public class OidcTenantConfig extends OidcClientCommonConfig implements io.quark
 
         public boolean enabled;
         public Optional<String> resource = Optional.empty();
+        public Optional<Set<String>> scopes = Optional.empty();
         public Optional<String> authorizationServer = Optional.empty();
         public boolean forceHttpsScheme = true;
 
@@ -2770,6 +2771,11 @@ public class OidcTenantConfig extends OidcClientCommonConfig implements io.quark
         @Override
         public Optional<String> resource() {
             return resource;
+        }
+
+        @Override
+        public Optional<Set<String>> scopes() {
+            return scopes;
         }
 
         @Override
@@ -2785,6 +2791,7 @@ public class OidcTenantConfig extends OidcClientCommonConfig implements io.quark
         private void addConfigMappingValues(io.quarkus.oidc.runtime.OidcTenantConfig.ResourceMetadata mapping) {
             enabled = mapping.enabled();
             resource = mapping.resource();
+            scopes = mapping.scopes();
             authorizationServer = mapping.authorizationServer();
             forceHttpsScheme = mapping.forceHttpsScheme();
         }

--- a/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java
+++ b/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java
@@ -865,13 +865,14 @@ public final class OidcTenantConfigBuilder extends OidcClientCommonConfigBuilder
      */
     public static final class ResourceMetadataBuilder {
 
-        private record ResourceMetadataImpl(boolean enabled, Optional<String> resource,
+        private record ResourceMetadataImpl(boolean enabled, Optional<String> resource, Optional<Set<String>> scopes,
                 Optional<String> authorizationServer, boolean forceHttpsScheme) implements ResourceMetadata {
         }
 
         private final OidcTenantConfigBuilder builder;
         private boolean enabled;
         private Optional<String> resource;
+        private Optional<Set<String>> scopes;
         private Optional<String> authorizationServer;
         private boolean forceHttpsScheme;
 
@@ -883,6 +884,8 @@ public final class OidcTenantConfigBuilder extends OidcClientCommonConfigBuilder
             this.builder = Objects.requireNonNull(builder);
             this.enabled = builder.resourceMetadata.enabled();
             this.resource = builder.resourceMetadata.resource();
+            this.scopes = builder.resourceMetadata.scopes().isEmpty() ? Optional.empty()
+                    : Optional.of(Set.copyOf(builder.resourceMetadata.scopes().get()));
             this.authorizationServer = builder.resourceMetadata.authorizationServer();
             this.forceHttpsScheme = builder.resourceMetadata.forceHttpsScheme();
         }
@@ -911,6 +914,23 @@ public final class OidcTenantConfigBuilder extends OidcClientCommonConfigBuilder
          */
         public ResourceMetadataBuilder resource(String resource) {
             this.resource = Optional.ofNullable(resource);
+            return this;
+        }
+
+        /**
+         * @param scopes {@link ResourceMetadata#scopes()}
+         * @return this builder
+         */
+        public ResourceMetadataBuilder scopes(String scope) {
+            return this.scopes(Set.of(scope));
+        }
+
+        /**
+         * @param scopes {@link ResourceMetadata#scopes()}
+         * @return this builder
+         */
+        public ResourceMetadataBuilder scopes(Set<String> scopes) {
+            this.scopes = Optional.ofNullable(scopes);
             return this;
         }
 
@@ -952,7 +972,7 @@ public final class OidcTenantConfigBuilder extends OidcClientCommonConfigBuilder
          * @return built ResourceMetadata
          */
         public ResourceMetadata build() {
-            return new ResourceMetadataImpl(enabled, resource, authorizationServer, forceHttpsScheme);
+            return new ResourceMetadataImpl(enabled, resource, scopes, authorizationServer, forceHttpsScheme);
         }
     }
 

--- a/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcTenantConfig.java
+++ b/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcTenantConfig.java
@@ -163,6 +163,11 @@ public interface OidcTenantConfig extends OidcClientCommonConfig {
         Optional<String> resource();
 
         /**
+         * Supported scopes.
+         */
+        Optional<Set<String>> scopes();
+
+        /**
          * Authorization server URL.
          * 'quarkus.oidc.auth-server-url' property value is reported by default.
          */

--- a/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/ResourceMetadataHandler.java
+++ b/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/ResourceMetadataHandler.java
@@ -146,6 +146,14 @@ public class ResourceMetadataHandler implements Handler<RoutingContext> {
             String resourceIdentifier = buildResourceIdentifierUrl(context, resolver, oidcConfig);
             metadata.put(OidcConstants.RESOURCE_METADATA_RESOURCE, resourceIdentifier);
 
+            if (oidcConfig.resourceMetadata().scopes().isPresent()) {
+                JsonArray scopes = new JsonArray();
+                for (String scope : oidcConfig.resourceMetadata().scopes().get()) {
+                    scopes.add(scope);
+                }
+                metadata.put(OidcConstants.RESOURCE_METADATA_SCOPES, scopes);
+            }
+
             JsonArray authorizationServers = new JsonArray();
             authorizationServers.add(0, oidcConfig.resourceMetadata().authorizationServer()
                     .orElse(oidcConfig.authServerUrl().get()));

--- a/extensions/oidc/runtime/src/test/java/io/quarkus/oidc/runtime/OidcTenantConfigImpl.java
+++ b/extensions/oidc/runtime/src/test/java/io/quarkus/oidc/runtime/OidcTenantConfigImpl.java
@@ -163,6 +163,7 @@ final class OidcTenantConfigImpl implements OidcTenantConfig {
         CERTIFICATION_CHAIN_TRUST_STORE_FILE_TYPE,
         RESOURCE_METADATA_ENABLED,
         RESOURCE_METADATA_RESOURCE,
+        RESOURCE_METADATA_SCOPES,
         RESOURCE_METADATA_AUTHORIZATION_SERVER,
         RESOURCE_METADATA_FORCE_HTTPS_SCHEME,
         LOGOUT_PATH,
@@ -620,6 +621,12 @@ final class OidcTenantConfigImpl implements OidcTenantConfig {
             @Override
             public Optional<String> resource() {
                 invocationsRecorder.put(ConfigMappingMethods.RESOURCE_METADATA_RESOURCE, true);
+                return Optional.empty();
+            }
+
+            @Override
+            public Optional<Set<String>> scopes() {
+                invocationsRecorder.put(ConfigMappingMethods.RESOURCE_METADATA_SCOPES, true);
                 return Optional.empty();
             }
 

--- a/integration-tests/oidc-code-flow/src/main/resources/application.properties
+++ b/integration-tests/oidc-code-flow/src/main/resources/application.properties
@@ -112,6 +112,7 @@ quarkus.oidc.tenant-refresh.token.refresh-expired=true
 quarkus.oidc.tenant-refresh.authentication.remove-redirect-parameters=false
 quarkus.oidc.tenant-refresh.authentication.cache-control=no-store
 quarkus.oidc.tenant-refresh.resource-metadata.enabled=true
+quarkus.oidc.tenant-refresh.resource-metadata.scopes=read,write
 
 quarkus.oidc.tenant-autorefresh.auth-server-url=${quarkus.oidc.auth-server-url}
 quarkus.oidc.tenant-autorefresh.client-id=quarkus-app
@@ -151,6 +152,7 @@ quarkus.oidc.tenant-nonce.token-state-manager.encryption-required=false
 quarkus.oidc.tenant-nonce.token.allow-jwt-introspection=false
 quarkus.oidc.tenant-nonce.resource-metadata.enabled=true
 quarkus.oidc.tenant-nonce.resource-metadata.resource=/metadata
+quarkus.oidc.tenant-nonce.resource-metadata.scopes=read
 quarkus.oidc.tenant-nonce.resource-metadata.authorization-server=http://localhost:8080/q/oidc
 
 quarkus.oidc.tenant-javascript.auth-server-url=${quarkus.oidc.auth-server-url}

--- a/integration-tests/oidc-code-flow/src/test/java/io/quarkus/it/keycloak/CodeFlowTest.java
+++ b/integration-tests/oidc-code-flow/src/test/java/io/quarkus/it/keycloak/CodeFlowTest.java
@@ -18,9 +18,11 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Base64;
 import java.util.List;
+import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.Callable;
 import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
 
 import javax.crypto.SecretKey;
 import javax.crypto.spec.SecretKeySpec;
@@ -142,7 +144,7 @@ public class CodeFlowTest {
             checkHealth();
 
             // Static default tenant
-            checkResourceMetadata(null, "/realms/quarkus");
+            checkResourceMetadata(null, "/realms/quarkus", null);
         }
     }
 
@@ -498,7 +500,7 @@ public class CodeFlowTest {
             assertEquals("Unexpected 401", ex.getMessage());
         }
         // Static `tenant-nonce` tenant with custom resource path
-        checkResourceMetadata("metadata", ":8080/q/oidc");
+        checkResourceMetadata("metadata", ":8080/q/oidc", Set.of("read"));
     }
 
     private void doTestCodeFlowNonce(boolean wrongRedirect) throws Exception {
@@ -918,7 +920,8 @@ public class CodeFlowTest {
             webClient.getCookieManager().clearCookies();
 
             // Static `tenant-refresh` tenant
-            checkResourceMetadata("tenant-refresh", "/realms/logout-realm");
+            checkResourceMetadata("tenant-refresh", "/realms/logout-realm", Set.of("read", "write"));
+
         }
     }
 
@@ -1844,7 +1847,7 @@ public class CodeFlowTest {
         return sessionCookie.getValue().split("\\|")[0];
     }
 
-    private static void checkResourceMetadata(String resource, String authorizationServerSuffix) {
+    private static void checkResourceMetadata(String resource, String authorizationServerSuffix, Set<String> scopes) {
         Response metadataResponse = RestAssured.when()
                 .get("http://localhost:8081" + OidcConstants.RESOURCE_METADATA_WELL_KNOWN_PATH
                         + (resource == null ? "" : "/" + resource));
@@ -1857,5 +1860,11 @@ public class CodeFlowTest {
         String authorizationServer = jsonAuthorizarionServers.getString(0);
         assertTrue(authorizationServer.startsWith("http://localhost:"));
         assertTrue(authorizationServer.endsWith(authorizationServerSuffix));
+
+        if (scopes != null) {
+            JsonArray scopesArray = jsonMetadata.getJsonArray(OidcConstants.RESOURCE_METADATA_SCOPES);
+            assertEquals(scopes.size(), scopesArray.size());
+            assertEquals(scopes, scopesArray.stream().collect(Collectors.toSet()));
+        }
     }
 }

